### PR TITLE
Use auth expires_in to refresh oauth tokens

### LINKED ISSUE #160: 3000 DNS queries a minute
After upgrading to 0.2.8 of ring-doorbell (in order to get the Oauth2 authentication, I'm now seeing about 3000 DNS queries, per minute, to oauth.ring.com.

### PR DESCRIPTION
**Pull request by jeromelaban**
Adds the use of `expires_in` from the oauth token to avoid getting new tokens too often.

Fixes #160

I'm not sure if refreshing at the exact time is the proper way to do it, maybe adding a 10s offset would avoid refresh issues.

### CODE REVIEW

File: ring_doorbell/__init__.py [hunk 1]
```diff
@@ -44,7 +45,6 @@ def __init__(self, username, password,
         self.params = None
         self._persist_token = persist_token
         self._push_token_notify_url = push_token_notify_url
-        self._timeout = timeout
```
Comment (tchellomello) [inline ring_doorbell/__init__.py]: Please, restore this attribute as this is being used as a timeout for the requests (see https://github.com/tchellomello/python-ring-doorbell/pull/167/files#diff-98d24069e9865c04cb18bcad289e3881L139 for example)

**Comment by tchellomello** [comment]
@jeromelaban in within 1hr, how considerable the DNS query changed using this new patch?

File: ring_doorbell/__init__.py (hunk 1 above)
Comment (jeromelaban) [reply ring_doorbell/__init__.py]: Apologies, I made the original changes on the stable branch, I missed that one :)

**Comment by jeromelaban** [comment]
> Hello @jeromelaban, thanks for your contribution.
> 
> Thank you for working on this!!

My pleasure :)

> @jeromelaban in within 1hr, how considerable the DNS query changed using this new patch?

From what I've seen, there's one resolution at startup, then one resolution per hour for `oauth.ring.com`, the rest is other ring.com hostnames. It also reduces the oauth queries to the same amount.

**Comment by steve-gombos** [comment]
So this works without changing too much.

The long term change for this is to move everything auth related into the auth class.  So the ring class won't have to care about if the token is expired or anything.  I haven't had time to look into this, and is also a fairly large change to this library, but the high level is:

- Ring class
  - Remove all username/password logic from Ring class
  - Pass auth instance into Ring constructor.  Caller takes responsibility for initializing/passing this in
  - Remove get_oauth_token method
  - Update query method to call Auth request method (this method will handle refreshing tokens if they are expired)
- Auth class 
  - Add cache file for caching the token responses, this will be processed to load existing tokens or generate a token if there isn't yet.
  - Update the request method to make sure it works with PUT, POST, GET HTTP request types.

Not sure if this makes sense.

**Comment by tchellomello** [comment]
Awesome @jeromelaban  @steve-gombos  Let's work together on these changes to move the `Auth` class. 

Good to be merged!!

File: ring_doorbell/__init__.py [hunk 2]
```diff
@@ -132,7 +144,6 @@ def _authenticate(self, attempts=RETRY_TOKEN, session=None, wait=1.0):
             loop += 1
 
             try:
-                if session is None:
```
Comment (d0ugal) [inline ring_doorbell/__init__.py]: The intention here looks wrong now.

Comment (tchellomello) [reply ring_doorbell/__init__.py]: @d0ugal, it was fixed by the PR #168 

Thanks for the heads up!

Comment (jeromelaban) [reply ring_doorbell/__init__.py]: Ah well! This should not have been there as well, my apologies. Thanks @d0ugal!

**Pull request merged**